### ISSUE TITLE
hs.uielement doesn't work with windows that have their role set to anything other than NSAccessibilityWindowRole

### ISSUE THREAD

**Issue by marshallbrekka** [Hammerspoon/hammerspoon#207]
Currently Emacs > 24.2 defines the window role as NSAccessibilityTextFieldRole (https://github.com/emacs-mirror/emacs/commit/c4328746b#diff-f43249227e5879c3fc3d7c08b1c41b1cR6048), which causes an exception when starting a watcher on the window because `push_element` doesn't create a window object.

One easy solution is to instead check for a property that only windows can have to determine if the element should be a window instead of an application or a generic uielement.

I have a patch that accomplishes that, but I wasn't sure if that would be the desired path to go down.

**Comment by cmsj** [comment]
Interesting. Let's see what @tmandry thinks - he wrote that extension.
Can you push up the patch somewhere? (A gist or pastebin maybe)

**Comment by marshallbrekka** [comment]
```objective-c
static void push_element(lua_State* L, AXUIElementRef element) {
    NSString* role = get_prop(element, NSAccessibilityRoleAttribute, @"");

    // The role attribute on a window can potentially be something
    // other than kAXWindowRole, so instead attempt to get an
    // attribute that is only defined on windows to determine if the
    // element is a window or not.
    NSNumber* isWindow = get_prop(element, NSAccessibilityMinimizedAttribute, nil);
    if (isWindow) {
        new_window(L, (AXUIElementRef)CFRetain(element));
    } else if ([role isEqualToString: (NSString*)kAXApplicationRole]) {
        pid_t pid;
        AXUIElementGetPid(element, &pid);
        new_application(L, pid);
    } else {
        new_uielement(L, (AXUIElementRef)CFRetain(element));
    }
}

```

**Comment by cmsj** [comment]
(fwiw, I think that duck-typing things makes sense, but I'd like to get a little more opinion on whether this will cause unintended side effects)

**Comment by tmandry** [comment]
This makes sense, though I'm not exactly an expert on the AX API (which is probably the reason for this fix :). Make sure the isWindow() method on uielement works as well.

**Comment by cmsj** [comment]
@marshallbrekka can you report what win:role() reports for an Emacs window please? I'm guessing it's not going to be AXWindow, so uielement.isWindow() will also need some work.

**Comment by cmsj** [comment]
@marshallbrekka that patch seems to work. I modified it slightly, to sell compare role to kAXWindowRole, so we're only testing the minimized attribute in the case of "unusual" windows. Felt a bit safer :)

We still need to test out win:role(), which I suspect will also cause hs.uielement:isWindow() to fail for things like Emacs.

**Comment by marshallbrekka** [comment]
personally I would think it would be desired to leave win:role() as is (seeing as its essentially a direct wrapper around the native call), but patching isWindow() would seem reasonable to me

**Comment by marshallbrekka** [comment]
I've put together a patch that should resole the isWindow issue, while still retaining the behavior of win:role()

https://github.com/Hammerspoon/hammerspoon/pull/216

any/all criticism is welcome :)

**Comment by cmsj** [comment]
#216 is now landed, so I think we can close this issue. thanks everyone :)

**Comment by tmandry** [comment]
This new `isWindow()` method is returning true for some AXApplications for me (notably Finder), which causes a bunch of problems. Maybe duck typing is not the answer. Maybe we just need a list of possible window roles.

**Comment by tmandry** [comment]
Since it's a NSAccessibilityTextFieldRole, we could check for a combination of that and a minimized attribute. It's a very specific workaround, so up to project maintainers whether to support it. But current implementation is broken :/

**Comment by cmsj** [comment]
@tmandry could you post a lua snippet showing the broken behaviour please?

**Comment by tmandry** [comment]
Yep, the problems occurred while using watchers. Here's a relatively simple snippet to reproduce:

```lua
local finder = hs.application.applicationsForBundleID('com.apple.finder')[1]

local watcher = finder:newWatcher(function(element, event)
  print("Finder application activated, role: "..element:role())
  print("isApplication: "..tostring(element:isApplication()))
  print("isWindow: "..tostring(element:isWindow()))

  local mainWindow = element:mainWindow()
  if mainWindow then
    print("Main window: "..mainWindow:title())
  end
end)

watcher:start({hs.uielement.watcher.applicationActivated})
```

Stick this into your init.lua and then focus a Finder window (or the desktop). Here's what I get:

```
Finder application activated, role: AXApplication
isApplication: true
isWindow: true
/Users/tyler/.hammerspoon/init.lua:59: attempt to call method 'mainWindow' (a nil value)
stack traceback:
	/Users/tyler/.hammerspoon/init.lua:59: in function 'callback'
	....app/Contents/Resources/extensions/hs/uielement/init.lua:105: in function 'handleEvent'
	....app/Contents/Resources/extensions/hs/uielement/init.lua:122: in function <....app/Contents/Resources/extensions/hs/uielement/init.lua:122>
```

Note that both `isApplication` and `isWindow` are true. Because the code thinks it has a window, it creates that kind of object and so `mainWindow` is not available. But the role is clearly `AXApplication`.

Do you want me to open a new issue?

**Comment by cmsj** [comment]
so, we could return false in isWindow() if the role is AXApplication. We're tying ourselves in ever tighter knots with this stuff, but that check at least feels like a fairly safe/sane thing to be doing. Any thoughts?

**Comment by tmandry** [comment]
My thinking is that it's better to make this fix only as broad as it needs to be, i.e. add an extra check for `NSAccessibilityTextFieldRole` elements that support minimization, rather than starting broad and fixing things that we find it breaks.

If it turns out this is an issue in many more apps than just Emacs, then maybe a broad approach is warranted, but otherwise I feel like it isn't worth risking the hard-to-find problems that could result.
